packages/swingset-liveslots/test/gc-before-finalizer.test.js

```diff
@@ -0,0 +1,228 @@
+  // examine the vref.
+
+  // This test simulates this case with our mockGC tools. It would
+  // fail if boyd-gc.js used getValForSlot instead of slotToVal.has.
+
+  // The GC code used to call getValForSlot() instead of
+  // slotToVal.has(), in the possiblyRetiredSet. The second test
+  // exercises this case, and would fail if it still used
+  // getValForSlot
+
+  let myPresence;
+  function buildRootObject(vatPowers) {
+    const { VatData } = vatPowers;
+    const { makeScalarBigMapStore } = VatData;
+    const store = makeScalarBigMapStore();
+    return Far('root', {
+      store: p => {
+        myPresence = p;
+        store.init('presence', p);
+      },
+      drop: () => store.delete('presence'),
+    });
+  }
+
+  const makeNS = () => ({ buildRootObject });
+  const ls = makeLiveSlots(syscall, 'vatA', {}, {}, gcTools, undefined, makeNS);
+  const { dispatch } = ls;
+  await dispatch(makeStartVat(kser()));
+
+  await dispatch(makeMessage('o+0', 'store', [kslot('o-1')]));
+  t.truthy(myPresence);
+
+  // clear out everything before our check
+  await dispatch(makeBringOutYourDead());
+  log.length = 0;
+
+  // drop the vdata reference
+  await dispatch(makeMessage('o+0', 'drop', []));
+  log.length = 0;
+
+  // and, before BOYD can happen, collect (but do not finalize) the Presence
+  gcTools.kill(myPresence);
+
+  // now BOYD
+  await dispatch(makeBringOutYourDead());
+
+  // the BOYD must not have done dropImports or retireImports
+  t.deepEqual(log, []);
+
+  // eventually, the finalizer runs
+  gcTools.flushAllFRs();
+
+  // *now* a BOYD will drop+retire
+  await dispatch(makeBringOutYourDead());
+  t.deepEqual(justGC(log), [
+    { type: 'dropImports', slots: ['o-1'] },
+    { type: 'retireImports', slots: ['o-1'] },
+  ]);
+});
+
+// disabled until #9956 and #9959 are fixed, which interfere with this
+// test
+
+/*
+
+test.failing('presence in COLLECTED state is not retired early', async t => {
+  const { syscall, log } = buildSyscall();
+  const gcTools = makeMockGC();
+
+  // The GC code used to call getValForSlot() instead of
+  // slotToVal.has(), in the possiblyRetiredSet. This test would fail
+  // if it still used getValForSlot.
+
+  // The setup is a Presence in the COLLECTED state as the only
+  // pillar, but not in possiblyDeadSet, whose vref also appears in
+  // possiblyRetiredSet (because a recognizer was just deleted). On
+  // the BOYD that handles the possiblyRetiredSet, the "reachability
+  // inhibits retirement" check should treat the COLLECTED state as
+  // reachable, so the retirement is deferred for a later BOYD (which
+  // would drop the vref first)
+  //
+  // To build this, we have an anchored (virtual) MapStore A holding
+  // the only reference (vdata) to a (virtual) WeakSetStore B. B has
+  // one (weak) key, o-1, for which there is a Presence P.
+  //
+  // 1: construct everything, kill the B Representative, BOYD. That
+  //    will process B, but leave it alive because of the vdata in A
```

Comment:
"A" and "B" are somewhat confusing here as my brain associates them with vats. Maybe "MS" and "WSS"?